exchanges/kucoin/kucoin_wrapper.go

```diff
@@ -0,0 +1,1460 @@
+		ku.SetEnabled(false)
+		return nil
+	}
+	err = ku.SetupDefaults(exch)
+	if err != nil {
+		return err
+	}
+
+	wsRunningEndpoint, err := ku.API.Endpoints.GetURL(exchange.WebsocketSpot)
+	if err != nil {
+		return err
+	}
+	err = ku.Websocket.Setup(
+		&stream.WebsocketSetup{
+			ExchangeConfig:        exch,
+			DefaultURL:            kucoinWebsocketURL,
+			RunningURL:            wsRunningEndpoint,
+			Connector:             ku.WsConnect,
+			Subscriber:            ku.Subscribe,
+			Unsubscriber:          ku.Unsubscribe,
+			GenerateSubscriptions: ku.GenerateDefaultSubscriptions,
+			Features:              &ku.Features.Supports.WebsocketCapabilities,
+			OrderbookBufferConfig: buffer.Config{
+				SortBuffer:            true,
+				SortBufferByUpdateIDs: true,
+				UpdateIDProgression:   true,
+			},
+			TradeFeed: ku.Features.Enabled.TradeFeed,
+		})
+	if err != nil {
+		return err
+	}
+	return ku.Websocket.SetupNewConnection(stream.ConnectionSetup{
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     exch.WebsocketResponseMaxLimit,
+		RateLimit:            500,
+	})
+}
+
+// Start starts the Kucoin go routine
+func (ku *Kucoin) Start(ctx context.Context, wg *sync.WaitGroup) error {
+	if wg == nil {
+		return fmt.Errorf("%T %w", wg, common.ErrNilPointer)
+	}
+	wg.Add(1)
+	go func() {
+		ku.Run(ctx)
+		wg.Done()
+	}()
+	return nil
+}
+
+// Run implements the Kucoin wrapper
+func (ku *Kucoin) Run(ctx context.Context) {
+	if ku.Verbose {
+		log.Debugf(log.ExchangeSys,
+			"%s Websocket: %s.",
+			ku.Name,
+			common.IsEnabled(ku.Websocket.IsEnabled()))
+		ku.PrintEnabledPairs()
+	}
+
+	if !ku.GetEnabledFeatures().AutoPairUpdates {
+		return
+	}
+
+	err := ku.UpdateTradablePairs(ctx, true)
+	if err != nil {
+		log.Errorf(log.ExchangeSys,
+			"%s failed to update tradable pairs. Err: %s",
+			ku.Name,
+			err)
+	}
+}
+
+// FetchTradablePairs returns a list of the exchanges tradable pairs
+func (ku *Kucoin) FetchTradablePairs(ctx context.Context, assetType asset.Item) (currency.Pairs, error) {
+	var cp currency.Pair
+	switch assetType {
+	case asset.Futures:
+		myPairs, err := ku.GetFuturesOpenContracts(ctx)
+		if err != nil {
+			return nil, err
+		}
+		pairs := make(currency.Pairs, 0, len(myPairs))
+		for x := range myPairs {
+			if !strings.EqualFold(myPairs[x].Status, "Open") {
+				continue
+			}
+			cp, err = currency.NewPairFromString(myPairs[x].Symbol)
+			if err != nil {
+				return nil, err
+			}
+			pairs = pairs.Add(cp)
+		}
+		configFormat, err := ku.GetPairFormat(asset.Futures, false)
+		if err != nil {
+			return nil, err
+		}
+		return pairs.Format(configFormat), nil
+	case asset.Spot, asset.Margin:
+		myPairs, err := ku.GetSymbols(ctx, "")
+		if err != nil {
+			return nil, err
+		}
+		pairs := make(currency.Pairs, 0, len(myPairs))
+		for x := range myPairs {
+			if !myPairs[x].EnableTrading {
+				continue
+			}
+			cp, err = currency.NewPairFromString(strings.ToUpper(myPairs[x].Name))
```

Comment:
Shazbert's suggestion made me look at this function a little more. I think it could be improved further. 

No `strings.Equalfold` which we've said before is expensive, no need to call `strings.ToUpper`,  no `ku.GetPairFormat` and `pairs.Format(configFormat)` and a quicker string index than string lookup
```go
		myPairs, err := ku.GetFuturesOpenContracts(ctx)
		if err != nil {
			return nil, err
		}
		pairs := make(currency.Pairs, 0, len(myPairs))
		for x := range myPairs {
			if strings.ToLower(myPairs[x].Status) != "open" {
				continue
			}
			cp, err = currency.NewPairFromStrings(myPairs[x].BaseCurrency, myPairs[x].Symbol[len(myPairs[x].BaseCurrency):])
			if err != nil {
				return nil, err
			}
			pairs = pairs.Add(cp)
		}
		return pairs, nil
	case asset.Spot, asset.Margin:
		myPairs, err := ku.GetSymbols(ctx, "")
		if err != nil {
			return nil, err
		}
		pairs := make(currency.Pairs, 0, len(myPairs))
		for x := range myPairs {
			if !myPairs[x].EnableTrading {
				continue
			}
			cp, err = currency.NewPairFromString(myPairs[x].Name)
```

Reply:
Then after that, please update config example and text configs

Reply:
I kept the underscore delimiter for the config has to have non empty string as a delimiter.

Reply:
Thank you Sam, but the config's have invalid pairs which was why I asked you to update them. We wish to keep the underscore delimiter, but please update the pairs